### PR TITLE
kubeadm: warning only for deprecated FG

### PR DESCRIPTION
**Pull request by pacoxu**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
- needed by https://github.com/kubernetes/kubernetes/pull/116570

#### Which issue(s) this PR fixes:

See https://github.com/kubernetes/kubeadm/issues/2346#issuecomment-1563079336
#### Special notes for your reviewer:

> BTW, there is no warning message for the GAed FG in kubeadm
#### Does this PR introduce a user-facing change?

```release-note
kubeadm: throw warnings instead of errors for deprecated feature gates
```

/cc @SataQiu  @neolit123

### CODE REVIEW

File: cmd/kubeadm/app/features/features.go [hunk 1]
```diff
@@ -148,7 +149,7 @@ func NewFeatureGate(f *FeatureList, value string) (map[string]bool, error) {
 		}
 
 		if featureSpec.PreRelease == featuregate.Deprecated {
-			return nil, errors.Errorf("feature-gate key is deprecated: %s", k)
+			klog.Warningf("Setting deprecated feature gate %s=%t. It will be removed in a future release.", k, v)
```
Comment (neolit123) [inline cmd/kubeadm/app/features/features.go:152]: i think @SataQiu pointed out we also error during upgrade. we might want to change that too.

Comment (pacoxu) [reply cmd/kubeadm/app/features/features.go:152]: https://github.com/kubernetes/kubernetes/blob/e5e354c73d06cf9071cd165bcf654f37fc437910/cmd/kubeadm/app/features/features.go#L150-L152

Here needs an update as well.

**Review by neolit123 — COMMENTED** [review]
/approve
/cc @SataQiu 

thanks!

**Comment by neolit123** [comment]
/release-note-edit

```release-note
kubeadm: throw warnings instead of errors for deprecated feature gates
```

**Comment by neolit123** [comment]
> BTW, there is no warning message for the GAed FG in kubeadm

we could show a warning. my preference is to be consistent with core k8s.

**Comment by SataQiu** [comment]
> > BTW, there is no warning message for the GAed FG in kubeadm
> 
> we could show a warning. my preference is to be consistent with core k8s.

I wonder if we can use feature-gate utils like other components?
For example 
https://github.com/kubernetes/kubernetes/blob/ef00ff21484336ec5207cee0c107acbe737b13af/cmd/kube-scheduler/app/server.go#L67-L70

Also we should update https://github.com/kubernetes/kubernetes/blob/ef00ff21484336ec5207cee0c107acbe737b13af/cmd/kubeadm/app/cmd/upgrade/common.go#L203-L209
 https://github.com/kubernetes/kubernetes/blob/ef00ff21484336ec5207cee0c107acbe737b13af/cmd/kubeadm/app/features/features.go#L167-L185

**Comment by pacoxu** [comment]
> we could log a separate k/kubeadm issue with some todos.

Agree. @SataQiu  we may open a cleanup task for the refactoring work.

**Comment by SataQiu** [comment]
@pacoxu thanks, overall LGTM.
Could you help to verify that the `kubeadm upgrade apply --feature-gates=UpgradeAddonsBeforeControlPlane=true` can work well with this patch?

**Comment by pacoxu** [comment]
> @pacoxu thanks, overall LGTM. Could you help to verify that the `kubeadm upgrade apply --feature-gates=UpgradeAddonsBeforeControlPlane=true` can work well with this patch?

Tested.
```
 ./kubeadm upgrade apply --feature-gates=UpgradeAddonsBeforeControlPlane=true v1.27.2
```

**Comment by pacoxu** [comment]
The warning message is like
```
W0529 13:35:47.552496   11534 features.go:152] Setting deprecated feature gate UpgradeAddonsBeforeControlPlane=%!t(string=true). It will be removed in a future release.
```
Not sure if it is expected.

**Comment by SataQiu** [comment]
> The warning message is like
> 
> > W0529 13:35:47.552496   11534 features.go:152] Setting deprecated feature gate UpgradeAddonsBeforeControlPlane=%!t(string=true). It will be removed in a future release.
> 
> Not sure if it is expected.
> 
> `%s=%t` -> `UpgradeAddonsBeforeControlPlane=%!t(string=true)`

Since it is the same as the [component-base](https://github.com/kubernetes/kubernetes/blob/d8e9a7b33a25e97b0880558fc94318a5d2fb3664/staging/src/k8s.io/component-base/featuregate/feature_gate.go#L238-L242) library, I think it is ok.
/lgtm
/approve

**Pull request merged**